### DISCUSSION TITLE
Callback Url in Next Auth

### DISCUSSION THREAD

**Discussion by PatrickNaenae** [vercel/next.js#55798] [Help]
### Summary

I deployed an app to Vercel, utilizing NextAuth for authentication. While everything functioned as expected in my local development environment, I encountered an issue when transitioning to production on Vercel. The problem lies in the fact that, upon attempting to sign in, the displayed callbackUrl points to localhost:3000. I'm puzzled as to why this local address is showing up as the callbackUrl in production mode. Despite having added NEXTAUTH_URL to the environment variables on Vercel, the issue persists. I've diligently attempted various solutions recommended to me, yet the problem remains unresolved. Below, I've attached screenshots for better understanding and also a link tot the github repo."

![Screen Shot 2023-09-22 at 07 30 23](https://github.com/vercel/next.js/assets/83951220/9ecad83b-e01d-4ad3-9552-fe31de60d114)
![Screen Shot 2023-09-22 at 07 30 54](https://github.com/vercel/next.js/assets/83951220/20ddd9e0-2cd1-4cba-a0ab-3fc92ddcc348)
![Screen Shot 2023-09-22 at 07 31 55](https://github.com/vercel/next.js/assets/83951220/183cf54a-e445-4a61-8d51-541f073241dd)
![Screen Shot 2023-09-22 at 07 32 03](https://github.com/vercel/next.js/assets/83951220/24be91b2-ed65-4ceb-adfc-75d060539eb0)
![Screen Shot 2023-09-22 at 07 32 46](https://github.com/vercel/next.js/assets/83951220/351e620d-1265-4d61-b3be-2f506f3d661f)
![Screen Shot 2023-09-22 at 07 33 02](https://github.com/vercel/next.js/assets/83951220/365581cc-0c41-41dd-a869-3fb7a3486d6f)
![Screen Shot 2023-09-22 at 07 33 13](https://github.com/vercel/next.js/assets/83951220/5dc5fd7e-27e5-4e17-a478-e21dd0efde78)

### Example

https://github.com/PatrickNaenae/dnd_gallery

**Comment by devjiwonchoi** [comment]
You don't need this:

```ts
const nextConfig = {
  env: {
    NEXTAUTH_URL: process.env.NEXTAUTH_URL || "http://localhost:3000", // will always be localhost:3000
  },
};
```

`env` is used to replace `process.env` and to expose on the client side.

Just use `process.env.NEXTAUTH_URL` directly with Vercel Environment Variable configured.

**Comment by koladee** [comment]
You didn't give enough details about the project, but I see you're trying to use the `signIn()` function from `next-auth` module.

> STEP 1

In your `signIn()` method, look for the `callbackUrl` parameter, be sure the assigned value is either an absolute URL `callbackUrl: 'https://my-next-app-example.vercel.app/<directory>'` or a relative URL `callbackUrl: '/<directory>'`

> STEP 2

Trace the file for the `/<directory>` route, look for `'use server'` in that file and any other file linked/imported into it, comment it out or remove it because any file containing `'use server'` is expected to only run in the server side and now you're trying to render/execute it on the client side.

> STEP 3

Push to your repository for vercel to sync the update if you have an active deployment of the project on vercel or deploy/redeploy the app on vercel if you haven't.


PS: To understand why removing `'use server'` will fix this issue, read up the use case of `'use server'`
